### PR TITLE
Supports update mode for FillNodes workload.

### PR DESCRIPTION
**Pull request by BrianSong**

### CODE REVIEW

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 1]
```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
...
 // Generates insert node.
 // For insert cases, the node's type will be `type` and its properties will be
 // generated w.r.t. `node_name`, `num_properties` and `string_value_bytes`.
 // Leaves it as return `void` instead of return `int64` for now because
 // FillNodes update mode will return `tensorflow::Status` in the future.
 template <typename T, typename N>
-void GenerateNodes(const NodesParam& nodes_param, const T& type,
-                   google::protobuf::RepeatedPtrField<N>& nodes,
-                   int64& curr_bytes) {
+tensorflow::Status GenerateNodes(const FillNodesConfig& fill_nodes_config,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc:321]: update docstring

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 2]
```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
...
+    }
+  }
+  // If the node's number of properties is greater than the type, uses custom
+  // properties instead.
+  while (populated_num_properties++ < nodes_param.num_properties) {
+    (*node.mutable_custom_properties())[absl::StrCat("custom_p-",
+                                                     populated_num_properties)]
+        .set_string_value(property_value);
+  }
+  return GetTransferredBytes<N>(node);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: move GetTransferredBytes to utils

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 3]
```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
...
+template <typename T, typename N>
+int64 SetNodePropertiesGivenType(const NodesParam& nodes_param, const T& type,
+                                 N& node) {
+  // Uses "********" as the fake property value for current node.
+  std::string property_value(nodes_param.string_value_bytes, '*');
+  // Loops over the types properties while generating the node's properties
+  // accordingly.
+  // TODO(briansong) Adds more property types support.
+  int64 populated_num_properties = 0;
+  for (const auto& p : type.properties()) {
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc:263]: write this in for condition

`++populated_num_properties > nodes_param.num_properties`

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 4]
```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
...
+                                        std::vector<Node>& existing_nodes,
+                                        N& node, Type& update_type) {
+  if (!existing_nodes.empty()) {
+    Node existing_node = existing_nodes.back();
+    existing_nodes.pop_back();
+    node = absl::get<N>(existing_node);
+    // node.set_type_id(absl::get<N>(existing_node).type_id());
+    // node.set_name(absl::get<N>(existing_node).name());
+    TF_RETURN_IF_ERROR(
+        SetTypeForUpdateNode(absl::get<N>(existing_node), store, update_type));
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: use, `SetTypeForUpdateNode(node, store, update_type));`

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 5]
```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
...
+}
+
+template <typename T, typename N>
+tensorflow::Status PrepareNodeForUpdate(const T& insert_type, const int64 i,
+                                        MetadataStore& store,
+                                        std::vector<Node>& existing_nodes,
+                                        N& node, Type& update_type) {
+  if (!existing_nodes.empty()) {
+    Node existing_node = existing_nodes.back();
+    existing_nodes.pop_back();
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: move this to line 185

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 6]
```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
...
+    request.add_type_ids(selected_node.type_id());
+    GetContextTypesByIDResponse response;
+    TF_RETURN_IF_ERROR(store.GetContextTypesByID(request, &response));
+    update_type = response.context_types(0);
+  }
+  return tensorflow::Status::OK();
+}
+
+template <typename N>
+tensorflow::Status InsertMakeUpNodeIntoDb(MetadataStore& store, N& node) {
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: `tensorflow::Status InsertMakeUpNodeIntoDb(MetadataStore& store, Artifact& node) `

`tensorflow::Status InsertMakeUpNodeIntoDb(MetadataStore& store, Execution& node) `

`tensorflow::Status InsertMakeUpNodeIntoDb(MetadataStore& store, Context& node) `

Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: s/InsertMakeUpNodeIntoDb/PrepareMakeUpNodeInDb

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 7]
```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
   return bytes;
 }
 
+template <typename N>
+tensorflow::Status SetTypeForUpdateNode(const N& selected_node,
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: ditto

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 8]
```diff
@@ -118,15 +119,116 @@ int64 GetTransferredBytes(const NT& node) {
...
+}
+
+template <typename T, typename N>
+int64 SetNodePropertiesGivenType(const NodesParam& nodes_param, const T& type,
+                                 N& node) {
+  // Uses "********" as the fake property value for current node.
+  std::string property_value(nodes_param.string_value_bytes, '*');
+  // Loops over the types properties while generating the node's properties
+  // accordingly.
+  // TODO(briansong) Adds more property types support.
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc:262]:
```
CP = MIN(|custom_properties|, nodes_param.num_properties) 

+ node.custom_properties clean |CP| 

RPC = remain_properties_to_change = (nodes_param.num_properties - CP)

```

1. for insert, no-op
2. for update, 
a) RPC = 0, return
b) RPC > 0, the same as the insert case

```
// for existing custom_properties, deleted entry
+ node.custom_properties clean |CP| 

if (rpc == 0) return;

// for existing properties, updated entry
for (const auto& p: type.properties()) {
    (*node..... ) = new value
    if (--rpc == 0) break;
}

// for custom properties, inserted new entry
while (rpc-- >= 0) {
   (*node...) custom properties = new value
}
```

**Review by hughmiao — CHANGES_REQUESTED** [review]
please finish the test, address the comment, and move get transfer bytes.

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc (hunk 2 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: Will address this in next PR. Thanks :)

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 9]
```diff
@@ -118,15 +119,177 @@ int64 GetTransferredBytes(const NT& node) {
...
 
-// Generates insert node.
-// For insert cases, the node's type will be `type` and its properties will be
-// generated w.r.t. `node_name`, `num_properties` and `string_value_bytes`.
-// Leaves it as return `void` instead of return `int64` for now because
-// FillNodes update mode will return `tensorflow::Status` in the future.
+// Sets ArtifactType `update_type` for modifying the selected update node's
+// properties in later SetNodePropertiesGivenType(). Returns detailed error if
+// query executions failed.
+template <typename N>
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: shall we drop this template N, and just use Artifact?

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 10]
```diff
@@ -118,15 +119,177 @@ int64 GetTransferredBytes(const NT& node) {
...
+  GetArtifactTypesByIDResponse response;
+  TF_RETURN_IF_ERROR(store.GetArtifactTypesByID(request, &response));
+  update_type = response.artifact_types(0);
+  return tensorflow::Status::OK();
+}
+
+// Sets ExecutionType `update_type` for modifying the selected update node's
+// properties in later SetNodePropertiesGivenType(). Returns detailed error if
+// query executions failed.
+template <typename N>
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: ditto

**Review by hughmiao — CHANGES_REQUESTED** [review]
is the CL ready to be reviewed?

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 11]
```diff
@@ -118,15 +119,174 @@ int64 GetTransferredBytes(const NT& node) {
...
+        (int64)node.custom_properties_size(), nodes_param.num_properties);
+    auto it = node.custom_properties().begin();
+    for (int64 i = 0; i < num_custom_properties_to_clear; ++i) {
+      node.mutable_custom_properties()->erase(it->first);
+      it++;
+    }
+  } else {
+    // For insert cases, this will be a no-op.
+    num_custom_properties_to_clear = 0;
+  }
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: drop this, and set = 0 in line 234

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc [hunk 12]
```diff
@@ -118,15 +119,174 @@ int64 GetTransferredBytes(const NT& node) {
...
+
+  // If there are no properties that needed to be changed further, return
+  // directly.
+  int64 remain_num_properties_to_change =
+      nodes_param.num_properties - num_custom_properties_to_clear;
+  if (remain_num_properties_to_change == 0) {
+    return GetTransferredBytes<N>(node);
```
Comment (hughmiao) [inline ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: should the transferred bytes be calculated differently for the update case?
we probably should count the ones that are deleted, and the properties that are changed or inserted, ignore the properties that are the same.

**Review by hughmiao — APPROVED** [review]
some comments left, please fix them before submit. 

Since the change only touched existing files and not conflict with the pending PR, let's submit a PR to the public repo after the fix.

File: ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc (hunk 12 above)
Comment (BrianSong) [reply ml_metadata/tools/mlmd_bench/fill_nodes_workload.cc]: Good point! Noted. I create two additional variable `transferred_bytes_for_cleared_node_properties` and `transferred_bytes_for_added_or_updated_node_properties` accordingly. PTAL.

**Comment by BrianSong** [comment]
Will create a separate public PR.

**Pull request closed without merging**